Реализовал методы Option

apavlychev opened the pull request:
Добавил реализацию для методов printIfAny, orElse, zip, filter, flatMap

apavlychev committed:
Реализовал методы Option

Добавил реализацию для методов printIfAny, orElse, zip, filter, flatMap

avoronets84 commented on src/main/scala/module1/02-fp.scala:
```diff
@@ -138,42 +138,94 @@ object hof{
+     * @tparam B - новый тип
+     * @return - возвращаемое значение
+     */
     def map[B](f: A => B): Option[B] = this match {
```
теперь когда у вас есть flatMap, map может выразить через него.

apavlychev replied:
Реализовал через flatMap в новом комите

avoronets84 commented on src/main/scala/module1/02-fp.scala:
```diff
@@ -138,42 +138,94 @@ object hof{
+     */
+    def printIfAny() = this match {
+      case Option.Some(v) => println(s"Значение: $v")
+      case Option.None => ()
```
не обязательно явно возвращать литерал для Unit

apavlychev replied:
Убрал () в новом комите

avoronets84 commented on src/main/scala/module1/02-fp.scala:
```diff
@@ -138,42 +138,94 @@ object hof{
+     *
+     * Реализовать метод zip, который будет создавать Option от пары значений из 2-х Option
+     */
+    def zip[B >: A](other: Option[B]): Option[(B, B)] = (this, other) match {
```
здесь нет необходимости указывать на отношения между типами, они могут быть совершенно разными

apavlychev replied:
Расширил тип B до свободного в новом комите

apavlychev committed:
Исправления после codereview

Расширил тип для zip, реализовал map через flatMap, убрал литерал () для printIfAny
